### PR TITLE
page:ProductListing ft. ProductFilterSort API done

### PR DESCRIPTION
**Pull request by manojsarna**
### Features Added 

**- [x] Pages - Products Listing**
**- [x] API - products API** 
**- [x] Functionality - Added Filters : Filter By Price, Rating, Categories, Brands , Out Of Stock**
**- [x] Functionality - Added Sort: Sort By Price, Sort By Rating**
**- [x] Functionality - Added Cart APIs**
**- [x] Features - Toast Container with Individual Toast Timeout Added** 
**- [x] Features - Mobile Responsive FilterSideBar** 

Deploy Link to Check : https://hopeful-newton-961563.netlify.app/

### Files To Check
[FilterSidebar.jsx](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-5918a0b7b5867dd140cc699e9ecec052fe035858b3d176beab1674512ee76936)
[/cart-context.js](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-b372b538d824a4879ea6eaf5c1879a1fe74a5128f71b912e04be9a462b2ac0e4)
[filterFunctions.js](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-9b1d6059200b37d17bfb7d2bce06796d104d56db0975be5f96174024a3254164)
[filterReducer.js](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-f715d2cae87b544e4e669120b8be222e4c29bec3300e4021ae7332cf91c4eabe)
[products-context.js](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-fce9acc5c15b554ed83544a1720ad1ccf4ef4d93126d5d9da53ad1ece1123d3a)
[Products.jsx](https://github.com/manojsarna/manoj-smash-cart-ecommerce-react/compare/productsListing-ProductsFiltering-API-functionality?expand=1#diff-aab6b7d220b66170c14cc3b283c5792f5c2e1be98c07a5b3fa8c91087856e3c5)
### Ignore Rest of the Files

image
image
image

### CODE REVIEW

File: src/context/products-context/filterReducer.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        searchString: action.payload,
+      };
+
+    case "FILTERS_RESET":
+      return initialState;
+
+    default:
+      throw new Error(`Unhandled type: ${action.type}`);
+  }
+};
```
Comment (b0llu) [inline src/context/products-context/filterReducer.js:90]: really liked the way how you handled all the cases in one file

**Review by b0llu — APPROVED** [review]
Really well written code Manoj. There is nothing to suggest you can merge

**Review by m26unkwn — APPROVED** [review]
Everything is good to go.

1. UI clean and beautiful
2.  Code is readable and clean.

File: src/context/cart-context/cart-context.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import { useContext, useState, createContext } from "react";
+import { useEffect } from "react";
```
Comment (SinchanaMS) [inline src/context/cart-context/cart-context.js:2]: You missed to call useEffect in line 1 itself.

File: src/components/card/Card.jsx [hunk 3]
```diff
@@ -18,22 +31,24 @@ export function Card({ product }) {
...
-            productInWishList && user
+            isProductInWishList && user
               ? "Remove From Wishlist"
               : "Add To Wishlist"
           }`}
           onClick={() => {
             if (!user) {
               navigate("/auth");
-            } else if (productInWishList) {
+            } else if (isProductInWishList) {
```
Comment (rahulyadav139) [inline src/components/card/Card.jsx:41]: This is a great state name with the prefix "is" specifying it as a boolean value. I like this prectice. great work.

**Review by SinchanaMS — APPROVED** [review]
Very neat code! Everything looks good!

File: src/components/products/filter-sidebar/BrandFilter.jsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            <input
+              type="checkbox"
+              name={brand.brand}
+              value={brand.brand}
+              checked={state.brands.includes(brand.brand)}
+              onChange={(e) => {
+                filterDispatch({
+                  type: "BRAND_FILTER",
+                  payload: e.target.value,
+                });
```
Comment (rahulyadav139) [inline src/components/products/filter-sidebar/BrandFilter.jsx]: brand.brand quite confusing. you can change it to understandable naming.

File: src/context/products-context/filterFunctions.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
+export const filterFunctions = (initialState, state, products) => {
+  //const { state, initialState, products } = useProducts();
+  console.log(initialState, state, products);
```
Comment (SinchanaMS) [inline src/context/products-context/filterFunctions.js:3]: Remove commented code and logs.

**Review by SinchanaMS — APPROVED** [review]

**Review by rahulyadav139 — APPROVED** [review]

File: src/context/products-context/filterReducer.js (hunk 1 above)
Comment (manojsarna) [reply src/context/products-context/filterReducer.js:90]: :)

File: src/context/cart-context/cart-context.js (hunk 2 above)
Comment (manojsarna) [reply src/context/cart-context/cart-context.js:2]: will check

File: src/context/products-context/filterFunctions.js (hunk 5 above)
Comment (manojsarna) [reply src/context/products-context/filterFunctions.js:3]: yup in refactoring

File: src/components/card/Card.jsx (hunk 3 above)
Comment (manojsarna) [reply src/components/card/Card.jsx:41]: Thanks 😇

File: src/components/products/filter-sidebar/BrandFilter.jsx (hunk 4 above)
Comment (manojsarna) [reply src/components/products/filter-sidebar/BrandFilter.jsx]: Done

**Pull request merged**